libs/mavlink/mavlink.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Phoenix-Pilot
+ *
+ * mavlink.c
+ *
+ * Mavlink protocol implementation
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-Pilot software
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <unistd.h>
+
+#include "mavlink.h"
+#include "mavlink_enums.h"
+
+
+#define MAVLINK_LABEL "mavlink"
+
+/* clang-format off */
+/* CRC_EXTRA of all payload types in `common.xml` dialect */
+#define MAVLINK_MESSAGE_CRCS {50, 124, 137, 0, 237, 217, 104, 119, 117, 0, 0, 89, 0, 0, 0, 0, 0, 0, 0, 0, 214, 159, 220, 168, 24, 23, 170, 144, 67, 115, 39, 246, 185, 104, 237, 244, 222, 212, 9, 254, 230, 28, 28, 132, 221, 232, 11, 153, 41, 39, 78, 196, 0, 0, 15, 3, 0, 0, 0, 0, 0, 167, 183, 119, 191, 118, 148, 21, 0, 243, 124, 0, 0, 38, 20, 158, 152, 143, 0, 0, 14, 106, 49, 22, 143, 140, 5, 150, 0, 231, 183, 63, 54, 47, 0, 0, 0, 0, 0, 0, 175, 102, 158, 208, 56, 93, 138, 108, 32, 185, 84, 34, 174, 124, 237, 4, 76, 128, 56, 116, 134, 237, 203, 250, 87, 203, 220, 25, 226, 46, 29, 223, 85, 6, 229, 203, 1, 195, 109, 168, 181, 47, 72, 131, 127, 0, 103, 154, 178, 200, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 189, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 36, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 208, 0, 0, 0, 0, 163, 105, 151, 35, 150, 179, 0, 0, 0, 0, 0, 90, 104, 85, 95, 130, 184, 81, 8, 204, 49, 170, 44, 83, 46, 0}
+/* clang-format on */
+
+#define MAVLINK_10_MAGIC      0xfe /* magic number for mavlink v1.0 */
+#define MAVLINK_20_MAGIC      0xfd /* magic number for mavlink v2.0 */
+#define MAVLINK_10_LEN_HEADER 6    /* header length for mavlink v1.0 */
+#define MAVLINK_20_LEN_HEADER 9    /* header length for mavlink v2.0 */
+#define MAVLINK_LEN_CHECKSUM  2    /* checksum field length */
+
+
+static const uint8_t crcExtra[] = MAVLINK_MESSAGE_CRCS;
+
+
+/* Mavlink CRC-16/MCRF4XX checksum calculator. Returns calculated checksum over `msgBuf` buffer of length `msgBufSz`. `msgBuf` must be a valid pointer. */
+static uint16_t mav_crc16(const uint8_t *msgBuf, uint16_t msgBufSz)
+{
+	uint16_t i = 0, crc = 0xffff;
+	uint8_t tmp;
```

Comment:
To make `<< 8` look less dangerous ;)
```suggestion
	uint16_t tmp;
```